File: Python-Übungen/teilbarkeits_regeln.py
def teilbar_durch_2(x):
    return int(x[-1]) % 2 == 0

def teilbar_durch_3(x):
    z = 0
    for i in range(len(x)):
        z += int(x[i])
    return z % 3 == 0

def teilbar_durch_4(x):
    y = x[-2:] if len(x) >= 2 else x
    return int(y) % 4 == 0

def teilbar_durch_5(x):
    y = int(x[-1])
    if y == 5 or y == 0:
        return True
    else:
        return False
    
def teilbar_durch_6():
    if teil_2 == True and teil_3 == True:
        return True
    else: 
        return False

def teilbar_durch_7(x):
    z = int(x)
    print(f"Starte mit: {z}")
    
    while abs(z) > 7:
        letzte = z % 10
        rest = z // 10
        z = rest - 2 * letzte
    if z == 0 or z == 7 or z == -7:
        return True
    else:
        return False


def teilbar_durch_8(x):
    y = x[-3:] if len(x) >= 2 else x
    return int(y) % 8 == 0

def teilbar_durch_9(x):
    z = 0
    for i in range(len(x)):
        z += int(x[i])
    return z % 9 == 0


while True:
    x = input("Gib eine Ganzzahl ein!\n")

    teil_2 = teilbar_durch_2(x)
    teil_3 = teilbar_durch_3(x)
    teil_4 = teilbar_durch_4(x)
    teil_5 = teilbar_durch_5(x)
    teil_6 = teilbar_durch_6()
    teil_7 = teilbar_durch_7(x)
    teil_8 = teilbar_durch_8(x)
    teil_9 = teilbar_durch_9(x)


    print("Teilbar durch 2:", teil_2)
    print("Teilbar durch 3:", teil_3)
    print("Teilbar durch 4:", teil_4)
    print("Teilbar durch 5:", teil_5)
    print("Teilbar durch 6:", teil_6)
    print("Teilbar durch 7:", teil_7)
    print("Teilbar durch 8:", teil_8)
    print("Teilbar durch 9:", teil_9)

File: C-Uebungen/teilbarkeitsregeln.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
int teilbar_durch_2(char x[50]) {
    int len = strlen(x);
    int digit = x[len - 1] - '0';
    return digit % 2 == 0;

}
int teilbar_durch_3(char x[50], int letters) {
    int y = 0;
    char* endptr;
    do{
      for (int i = 0; i < letters; i++) {
        y = y+(int)x[i];
    }  
    } while (y > 9);
    if (y==3 || y==6 || y==9) {
        return 1;
    }
    else{
        return 0;
    }
    
}
int teilbar_durch_4(char x[50]) {
    int len = strlen(x);
    char num[3];  // 2 characters + null terminator
    num[0] = x[len - 2];
    num[1] = x[len - 1];
    num[2] = '\0';  // Null terminator
    if (atoi(num) == 0) {
        return 1;
    }
    else {
        return 0;
    }
}
int teilbar_durch_5(char x[50]) {
    int len = strlen(x);
    int num;
    num = x[len - 1];
    if ((strlen(x)-1) == 0 || (int)x[-1] == 5) {
        return 1;
    }
    else {
        return 0;
    }
}
int teilbar_durch_6(char x[50], int letters) {
    if (teilbar_durch_2(x) == 1 && teilbar_durch_3(x, letters) == 1) {
        return 1;
    }
    else{
        return 0;
    }
}
int teilbar_durch_7(char x[50]) {
    int rest = atoi(x);
    int z = 0;
    do
    {
        z = rest%10;
        rest = (int)(rest/10) - (z*2);
    } while (rest > 9);
    if (rest == 0 || rest == 7) {
        return 1;
    }
    else{
        return 0;
    }
    

}
int teilbar_durch_8(char x[50]) {
    int len = strlen(x);
    char num[4];  // 2 characters + null terminator
    num[0] = x[len - 3];
    num[1] = x[len - 2];
    num[2] = x[len - 1];
    num[3] = '\0';  // Null terminator
    if ((atoi(num)%8) == 0) {
        return 1;
    }
    else {
        return 0;
    }
}
int teilbar_durch_9(char x[50], int letters) {
    int y = 0;
    char* endptr;
    do{
      for (int i = 0; i < letters; i++) {
        y = y+strlen(x);
    }  
    } while (y > 9);
    if (y==9) {
        return 1;
    }
    else{
        return 0;
    }
}

int main() {
    char x[50];
    printf("Bitte gib eine natürliche Zahl an!\n");
    scanf("%49s", x);
    printf("%s\n", x);
    int letters = strlen(x);

    

    //Teilbar durch 2?
    if (teilbar_durch_2(x) == 1) {
        printf("Ist teilbar durch 2\n");
    }
    else {
        printf("Ist nicht teilbar durch 2\n");
    }
    //Teilbar durch 3?
    if (teilbar_durch_3(x, letters) == 1) {
        printf("Ist teilbar durch 3\n");
    }
    else {
        printf("Ist nicht teilbar durch 3\n");
    }
    //Teilbar durch 4?
    if (teilbar_durch_4(x) == 1) {
        printf("Ist teilbar durch 4\n");
    }
    else {
        printf("Ist nicht teilbar durch 4\n");
    }
    //Teilbar durch 5?
    if (teilbar_durch_5(x) == 1) {
        printf("Ist teilbar durch 5\n");
    }
    else {
        printf("Ist nicht teilbar durch 5\n");
    }
    //Teilbar durch 6?
    if (teilbar_durch_6(x, letters) == 1) {
        printf("Ist teilbar durch 6\n");
    }
    else {
        printf("Ist nicht teilbar durch 6\n");
    }
    //Teilbar durch 7?
    if (teilbar_durch_7(x) == 1) {
        printf("Ist teilbar durch 7\n");
    }
    else {
        printf("Ist nicht teilbar durch 7\n");
    }
    //Teilbar durch 8?
    if (teilbar_durch_8(x) == 1) {
        printf("Ist teilbar durch 8\n");
    }
    else {
        printf("Ist nicht teilbar durch 8\n");
    }
    //Teilbar durch 9?
    if (teilbar_durch_9(x, letters) == 1) {
        printf("Ist teilbar durch 9\n");
    }
    else {
        printf("Ist nicht teilbar durch 9\n");
    }


}
